### PR TITLE
Hotfix/fused ce triton

### PR DESCRIPTION
**Pull request by sarthak-amd**
**Kernel fixes:**
- Compute valid token count on host: `denom = max(1, (target != ignore_idx).sum())`
- Scale loss and gradients by `denom` instead of `n_rows`
- Add `grad_output_stride` parameter, compute dynamically: `1 if grad_output.numel() > 1 else 0`
- Add `is_cg_capturable` flag for CUDA graph compatibility

**Test improvements:**
- Add explicit loss value assertions vs `torch.nn.CrossEntropyLoss`
- Use `torch.square(loss)` for non-trivial backward
- Apply dtype-aware tolerances

### CODE REVIEW

File: transformer_engine/pytorch/cross_entropy.py
Comment (ipanfilo) [inline transformer_engine/pytorch/cross_entropy.py:1]: What is a purpose of this file change?

File: transformer_engine/pytorch/triton/cross_entropy.py
Comment (ipanfilo) [inline transformer_engine/pytorch/triton/cross_entropy.py:1]: Modify copyright date. The file is also maintained in upstream so avoid unnecessary reformattings

File: transformer_engine/pytorch/triton/cross_entropy.py [hunk 1]
```diff
@@ -118,7 +117,7 @@ def cross_entropy_kernel(
     m_d_X_y_stride: The stride of m/d/X_y tensor.
     rank (int): The rank of this device in the TP group.
     world_size (int): The size of world involved in this distributed loss calculation.
-    ignore_idx (int): Tokens to be ignored for loss and gradient calculation.
+    ignore_idx (int): Tokens to be ignored for loss and gradient calculation. (default -100)
```
Comment (ipanfilo) [inline transformer_engine/pytorch/triton/cross_entropy.py:120]: There is no default here.

File: transformer_engine/pytorch/triton/cross_entropy.py [hunk 2]
```diff
@@ -334,25 +338,29 @@ def cross_entropy_forward(
...
 
-    loss = torch.reshape(loss_1d, (B, SQ)) if not reduce_loss else (torch.sum(loss_1d) / n_rows)
+    loss = torch.reshape(loss_1d, (B, SQ)) if not reduce_loss else (torch.sum(loss_1d) / denom)
 
     return loss, _input
 
 
-def cross_entropy_backward(_input: torch.Tensor, grad_output: torch.Tensor):
+def cross_entropy_backward(
+    _input: torch.Tensor, grad_output: torch.Tensor, is_cg_capturable: bool = False
```
Comment (ipanfilo) [inline transformer_engine/pytorch/triton/cross_entropy.py:354]: This code interferes and conflicts with upcoming IFU 2.8

**Review by ipanfilo — CHANGES_REQUESTED** [review]

**Review by wenchenvincent — CHANGES_REQUESTED** [review]
@sarthak-amd As mentioned in the previous PR, could you refactor the PR as 3 commits:

- 2 commits would be cherrypicking from the upstream PRs. (https://github.com/NVIDIA/TransformerEngine/pull/1879, https://github.com/NVIDIA/TransformerEngine/pull/2139)
- 1 commit for the ignore_idx with a test to cover it.
This way the PR would be very clear and easy to understand.

**Comment by ipanfilo** [comment]
> @sarthak-amd As mentioned in the previous PR, could you refactor the PR as 3 commits:
> 
>     * 2 commits would be cherrypicking from the upstream PRs. ([NVIDIA/TransformerEngine#1879](https://github.com/NVIDIA/TransformerEngine/pull/1879), [NVIDIA/TransformerEngine#2139](https://github.com/NVIDIA/TransformerEngine/pull/2139))
> 
>     * 1 commit for the ignore_idx with a test to cover it.
>       This way the PR would be very clear and easy to understand.

The one of aforementioned PRs is part of IFU 2.6, i.e. part of ROCm TE already, the other is part of IFU 2.8

**Comment by wenchenvincent** [comment]
@sarthak-amd With the dev synced to v2.8. I think we only need a commit with your own fix. Could you update the PR with that?
